### PR TITLE
feat(makeServerFetchye): adding run method to response

### PR DESCRIPTION
**Pull request by jjallwood**
We have been experimenting with using `fetchye` across our tenancy, and have run into an interesting case where we want to be able to use both the `data`, and the reload `run` methods as part of our module thunk methods. Currently we get both the `data` object and the `run` method into our components via the `useFetchye` hook, we then pass these into our redux thunks as parameters. However this method is clumsy and starts to get heavy when the example includes more than one source of data (we managed to get up to 3x in one case), or when thunks begin to chain calls.

This change is to export the `run` method, so both the two interfaces to Fetchye expose the same API to the client. This would allow us to re-use the underlying `makeOneServerFetchye` methods not only in our `loadModuleData` functions, but in redux thunks to load / get / reload data without getting stuck in parameter hell.

Since we have started this change the `oneFetchye` method was added to `fetchye-one-app` which compliments our change allowing engineers to just dispatch their fetchye requests in their thunk. 

## What we have today

```javascript
const onPrimaryTravelerSaveThunk = ({
  contactInfo, // fetchye run and data are renamed and passed as a param to the thunk
  reloadContactInfo, 
}) => async (dispatch, getState) => {
  const state = getState();
  await dispatch(setFormSubmitted(true));

  if (hasFormChanged(contactInfo, state)) {
    dispatch(setIsSaving());
    try {
      await dispatch(postUpdatesThunk());
      await dispatch(setSaveSuccessful());
    } catch (error) {
      await dispatch(setErrorSaving());
    } finally {
      reloadContactInfo();
    }
  }
  dispatch(setFormSubmitted(false));
};
```

## What we would like to be able to do

```javascript
const onPrimaryTravelerSaveThunk = () => async (dispatch, getState) => {
  const state = getState();
  await dispatch(setFormSubmitted(true));

  const {
    data: contactInfo,
    run: reloadContactInfo,
  } = await dispatch(loadContactInfo());  // by accessing data directly from fetchye via a thunk. we can remove the parameters

  if (hasFormChanged(contactInfo, state)) {
    dispatch(setIsSaving());
    try {
      await dispatch(postUpdatesThunk());
      await dispatch(setSaveSuccessful());
    } catch (error) {
      await dispatch(setErrorSaving());
    } finally {
      reloadContactInfo();
    }
  }
  dispatch(setFormSubmitted(false));
};
```

## Here is how we access the exported `data`, `errors` and `run` response within a thunk context via our `loadContactInfo` function

``` javascript
export const loadContactInfo = ({
  isCompanion,
  companionId,
}) => async (dispatch, getState) => {
  const { data: profileId } = await dispatch(loadProfileId);
  const { data: secToken } = await dispatch(loadSecToken);
  return dispatch(oneFetchye(
    getContactInfoUrl(profileId, isCompanion, companionId)(getState()),
    {
      headers: generateHeaders(secToken),
      credentials: 'include',
    }
  ));
};
```

### CODE REVIEW

File: packages/fetchye/src/useFetchye.js [hunk 1]
```diff
@@ -78,16 +75,12 @@ const useFetchye = (
...
     ),
     run() {
-      const runOptions = handleDynamicHeaders(options);
-      const runComputedKey = typeof options.headers === 'function'
-        ? computeKey(key, defaultMapOptionsToKey(mapOptionsToKey(runOptions)))
-        : computedKey;
       return runAsync({
         dispatch,
-        computedKey: runComputedKey,
+        computedKey,
```
Comment (code-forger) [inline packages/fetchye/src/useFetchye.js:80]:
```suggestion
        computedKey: computeKey(key, options),
```

I think you still need to re-compute the key inside the run function.

This allows the `handleDynamicHeaders` function to re-call the users dynamic header function, re-creating their dynamic data as close to the `run` as possible

File: packages/fetchye/src/makeServerFetchye.js [hunk 2]
```diff
@@ -27,18 +26,27 @@ const makeServerFetchye = ({
...
-  { mapOptionsToKey = (options) => options, ...options } = { },
+  options = {},
   fetcher = ssrFetcher
 ) => {
   const { cacheSelector } = cache;
-  const computedKey = computeKey(key, defaultMapOptionsToKey(mapOptionsToKey(options)));
+  const computedKey = computeKey(key, options);
+  const run = () => runAsync({
+    dispatch,
+    computedKey,
```
Comment (code-forger) [inline packages/fetchye/src/makeServerFetchye.js:36]:
```suggestion
    computedKey: computeKey(key, options),
```
Same as in useFetchye:

the dynamic headers need to be recreated inside `run`

File: packages/fetchye/src/makeServerFetchye.js [hunk 3]
```diff
@@ -27,18 +26,27 @@ const makeServerFetchye = ({
   fetchClient,
 }) => async (
   key,
-  { mapOptionsToKey = (options) => options, ...options } = { },
+  options = {},
   fetcher = ssrFetcher
 ) => {
   const { cacheSelector } = cache;
-  const computedKey = computeKey(key, defaultMapOptionsToKey(mapOptionsToKey(options)));
+  const computedKey = computeKey(key, options);
```
Comment (code-forger) [inline packages/fetchye/src/makeServerFetchye.js:33]: With the last suggestion, this line can then dissapear

File: packages/fetchye/__tests__/useFetchye.spec.jsx [hunk 4]
```diff
@@ -123,26 +124,33 @@ describe('useFetchye', () => {
...
+        expect(global.fetch).toHaveBeenCalledTimes(2);
+        expect(global.fetch).toHaveBeenNthCalledWith(1, 'http://example.com', {
+          headers: {
+            dynamicHeader: 'dynamic value 0',
+          },
+          mapOptionsToKey: expect.any(Function),
+        });
+        expect(global.fetch).toHaveBeenNthCalledWith(2, 'http://example.com', {
+          headers: {
+            dynamicHeader: 'dynamic value 1',
```
Comment (code-forger) [inline packages/fetchye/__tests__/useFetchye.spec.jsx:150]: 🚀

**Review by code-forger — APPROVED** [review]

**Review by Matthew-Mallimo — APPROVED** [review]

**Pull request merged**